<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Using codon models &#8212; cogent3  documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Use an empirical protein substitution model" href="empirical_protein_models.html" />
    <link rel="prev" title="Allowing substitution model parameters to differ between branches" href="scope_model_params_on_trees.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../install.html">Install</a></li>
                <li><a href="https://cogent3.org/docs">Docs</a></li>
                <li><a href="https://github.com/cogent3/cogent3">Code</a></li>
                <li><a href="https://cogent3.org/cogent3/draw/index.html">Gallery</a></li>
                <li><a href="../projects.html">Projects Using</a></li>
                <li><a href="https://github.com/cogent3/c3dev/wiki/Code-of-Conduct">Code of Conduct</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">The Library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Using codon models</a><ul>
<li><a class="reference internal" href="#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li><a class="reference internal" href="#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li><a class="reference internal" href="#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li><a class="reference internal" href="#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="scope_model_params_on_trees.html" title="Previous Chapter: Allowing substitution model parameters to differ between branches"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Allowing subs...</span>
    </a>
  </li>
  <li>
    <a href="empirical_protein_models.html" title="Next Chapter: Use an empirical protein substitution model"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Use an empiri... &raquo;</span>
    </a>
  </li>
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="using-codon-models">
<h1>Using codon models<a class="headerlink" href="#using-codon-models" title="Permalink to this headline">¶</a></h1>
<p>The basic paradigm for evolutionary modelling is:</p>
<ol class="arabic simple">
<li><p>construct the codon substitution model</p></li>
<li><p>constructing likelihood function</p></li>
<li><p>modify likelihood function (setting rules)</p></li>
<li><p>providing the alignment(s)</p></li>
<li><p>optimisation</p></li>
<li><p>get results out</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the following, a result followed by ‘…’ just means the output has been truncated for the sake of a succinct presentation.</p>
</div>
<div class="section" id="constructing-the-codon-substitution-model">
<h2>Constructing the codon substitution model<a class="headerlink" href="#constructing-the-codon-substitution-model" title="Permalink to this headline">¶</a></h2>
<p>Cogent3 implements 4 basic rate matrices, described in a recently accepted manuscript: i) NF models, these are nucleotide frequency weighted rate matrices and were initially described by Muse and Gaut (1994); ii) a variant of (i) where position specific nucleotide frequencies are used; iii) TF models, these are tuple (codon in this case) frequency weighted rate matrices and were initially described by Goldman and Yang (1994); iv) CNF, these use the conditional nucleotide frequency and have developed by Yap, Lindsay, Easteal and Huttley. These different models can be created using provided convenience functions which will be the case here, or specified by directly calling the <code class="docutils literal notranslate"><span class="pre">TimeReversibleCodon</span></code> substitution model class and setting the argument <code class="docutils literal notranslate"><span class="pre">mprob_model</span></code> equal to:</p>
<ul class="simple">
<li><p>NF, <code class="docutils literal notranslate"><span class="pre">mprob_model='monomer'</span></code></p></li>
<li><p>NF with position specific nucleotide frequencies, <code class="docutils literal notranslate"><span class="pre">mprob_model='monomers'</span></code></p></li>
<li><p>TF, <code class="docutils literal notranslate"><span class="pre">mprob_model=None</span></code></p></li>
<li><p>CNF, <code class="docutils literal notranslate"><span class="pre">mprob_model='conditional'</span></code></p></li>
</ul>
<p>In the following I will construct GTR variants of i and iv and a HKY variant of iii.</p>
<p>We import these explicitly from the <code class="docutils literal notranslate"><span class="pre">cogent3.evolve.models</span></code> module.</p>
<p>These are functions and calling them returns the indicated substitution model with default behaviour of recoding gap characters into N’s.</p>
<p>In the following demonstration I will use only the CNF form (<code class="docutils literal notranslate"><span class="pre">cnf</span></code>).</p>
<p>For our example we load a sample alignment and tree as per usual. To reduce the computational overhead for this example we will limit the number of sampled taxa.</p>
</div>
<div class="section" id="standard-test-of-neutrality">
<h2>Standard test of neutrality<a class="headerlink" href="#standard-test-of-neutrality" title="Permalink to this headline">¶</a></h2>
<p>We construct a likelihood function and constrain omega parameter (the ratio of nonsynonymous to synonymous substitutions) to equal 1. We also set some display formatting parameters.</p>
<p>We then provide an alignment and optimise the model. In the current case we just use the local optimiser (hiding progress to keep this document succinct). We then print(the results.)</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>I’m going to specify a set of conditions that will be used for all optimiser steps. For those new to python, one can construct a dictionary with the following form <code class="docutils literal notranslate"><span class="pre">{'argument_name':</span> <span class="pre">argument_value}</span></code>, or alternatively <code class="docutils literal notranslate"><span class="pre">dict(argument_name=argument_value)</span></code>. I’m doing the latter. This dictionary is then passed to functions/methods by prefacing it with <code class="docutils literal notranslate"><span class="pre">**</span></code>.</p>
</div>
<p>In the above output, the first table shows the maximum likelihood estimates (MLEs) for the substitution model parameters that are ‘global’ in scope. For instance, the <code class="docutils literal notranslate"><span class="pre">C/T=4.58</span></code> MLE indicates that the relative rate of substitutions between C and T is nearly 5 times the background substitution rate.</p>
<p>The above function has been fit using the default counting procedure for estimating the motif frequencies, i.e. codon frequencies are estimated as the average of the observed codon frequencies. If you wanted to numerically optimise the motif probabilities, then modify the likelihood function creation line to</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span> <span class="o">=</span> <span class="n">cnf</span><span class="o">.</span><span class="n">make_likelihood_function</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">optimise_motif_probs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>We can then free up the omega parameter, but before we do that we’ll store the log-likelihood and number of free parameters for the current model form for reuse later.</p>
<p>We then conduct a likelihood ratio test whether the MLE of omega significantly improves the fit over the constraint it equals 1. We import the convenience function from the <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> stats module.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.maths.stats</span> <span class="kn">import</span> <span class="n">chisqprob</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">LR</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">non_neutral_lnL</span> <span class="o">-</span> <span class="n">neutral_lnL</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">non_neutral_nfp</span> <span class="o">-</span> <span class="n">neutral_nfp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">chisqprob</span><span class="p">(</span><span class="n">LR</span><span class="p">,</span> <span class="n">df</span><span class="p">))</span>
<span class="go">0.0026...</span>
</pre></div>
</div>
<p>Not surprisingly, this is significant. We then ask whether the Human and Chimpanzee edges have a value of omega that is significantly different from the rest of the tree.</p>
<p>This is basically a replication of the original Huttley et al (2000) result for <em>BRCA1</em>.</p>
</div>
<div class="section" id="rate-heterogeneity-model-variants">
<h2>Rate-heterogeneity model variants<a class="headerlink" href="#rate-heterogeneity-model-variants" title="Permalink to this headline">¶</a></h2>
<p>It is also possible to specify rate-heterogeneity variants of these models. In the first instance we’ll create a likelihood function where these rate-classes are global across the entire tree. Because fitting these models can be time consuming I’m going to recreate the non-neutral likelihood function from above first, fit it, and then construct the rate-heterogeneity likelihood function. By doing this I can ensure that the richer model starts with parameter values that produce a log-likelihood the same as the null model, ensuring the subsequent optimisation step improves the likelihood over the null.</p>
<p>Now, we have a null model which we know (from having fit it above) has a MLE &lt; 1. We will construct a rate-heterogeneity model with just 2 rate-classes (neutral and adaptive) that are separated by the boundary of omega=1. These rate-classes are specified as discrete bins in Cogent3 and the model configuration steps for a bin or bins are done using the <code class="docutils literal notranslate"><span class="pre">set_param_rule</span></code> method. To ensure the alternate model starts with a likelihood at least as good as the previous we need to make the probability of the neutral site-class bin ~= 1 (these are referenced by the <code class="docutils literal notranslate"><span class="pre">bprobs</span></code> parameter type) and assign the null model omega MLE to this class.</p>
<p>To get all the parameter MLEs (branch lengths, GTR terms, etc ..) into the alternate model we get an annotated tree from the null model which will have these values associated with it.</p>
<p>We can then construct a new likelihood function, specifying the rate-class properties.</p>
<p>We define a very small value (<code class="docutils literal notranslate"><span class="pre">epsilon</span></code>) that is used to specify the starting values.</p>
<p>We now provide starting parameter values for <code class="docutils literal notranslate"><span class="pre">omega</span></code> for the two bins, setting the boundary</p>
<p>and provide the starting values for the bin probabilities (<code class="docutils literal notranslate"><span class="pre">bprobs</span></code>).</p>
<p>The above statement essentially assigns a probability of nearly 1 to the ‘neutral’ bin. We now set the alignment and fit the model.</p>
<p>We can get the posterior probabilities of site-classifications out of this model as</p>
<p>This is a <code class="docutils literal notranslate"><span class="pre">DictArray</span></code> class which stores the probabilities as a <code class="docutils literal notranslate"><span class="pre">numpy.array</span></code>.</p>
</div>
<div class="section" id="mixing-branch-and-site-heterogeneity">
<h2>Mixing branch and site-heterogeneity<a class="headerlink" href="#mixing-branch-and-site-heterogeneity" title="Permalink to this headline">¶</a></h2>
<p>The following implements a modification of the approach of Zhang, Nielsen and Yang (Mol Biol Evol, 22:2472–9, 2005). For this model class, there are groups of branches for which all positions are evolving neutrally but some proportion of those neutrally evolving sites change to adaptively evolving on so-called foreground edges. For the current example, we’ll define the Chimpanzee and Human branches as foreground and everything else as background. The following table defines the parameter scopes.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 19%" />
<col style="width: 22%" />
<col style="width: 30%" />
<col style="width: 29%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Site class</p></th>
<th class="head"><p>Proportion</p></th>
<th class="head"><p>Background edges</p></th>
<th class="head"><p>Foreground edges</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0</p></td>
<td><p>p_0</p></td>
<td><p>0 &lt; omega_0 &lt; 1</p></td>
<td><p>0 &lt; omega_0 &lt; 1</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>p_1</p></td>
<td><p>omega_1=1</p></td>
<td><p>omega_1=1</p></td>
</tr>
<tr class="row-even"><td><p>2a</p></td>
<td><p>p_2</p></td>
<td><p>0 &lt; omega_0 &lt; 1</p></td>
<td><p>omega_2 &gt; 1</p></td>
</tr>
<tr class="row-odd"><td><p>2b</p></td>
<td><p>p_3</p></td>
<td><p>omega_1=1</p></td>
<td><p>omega_2 &gt; 1</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Our implementation is not as parametrically succinct as that of Zhang et al, we have 1 additional bin probability.</p>
</div>
<p>After Zhang et al, we first define a null model that has 2 rate classes ‘0’ and ‘1’. We also get all the MLEs out using <code class="docutils literal notranslate"><span class="pre">get_statistics</span></code>, just printing out the bin parameters table in the current case.</p>
<p>We’re also going to use the MLEs from the <code class="docutils literal notranslate"><span class="pre">rate_lf</span></code> model, since that nests within the more complex branch by rate-class model. This is unfortunately quite ugly compared with just using the annotated tree approach described above. It is currently necessary, however, due to a bug in constructing annotated trees for models with binned parameters.</p>
<p>We now create the more complex model,</p>
<p>and set from the nested null model the branch lengths,</p>
<p>GTR term MLES,</p>
<p>binned parameter values,</p>
<p>and the bin probabilities.</p>
<p>The result of these steps is to create a rate/branch model with initial parameter values that result in likelihood the same as the null.</p>
<p>This function can then be optimised as before. The results of one such optimisation are shown below. As you can see, the <code class="docutils literal notranslate"><span class="pre">omega</span></code> value for the ‘2a’ and ‘2b’ bins is at the upper bounds, indicating the model is not maximised in this case.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rate_branch_lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="o">**</span><span class="n">optimiser_args</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rate_branch_lf</span><span class="p">)</span>
<span class="n">Likelihood</span> <span class="n">function</span> <span class="n">statistics</span>
<span class="n">log</span><span class="o">-</span><span class="n">likelihood</span> <span class="o">=</span> <span class="o">-</span><span class="mf">6753.4561</span>
<span class="n">number</span> <span class="n">of</span> <span class="n">free</span> <span class="n">parameters</span> <span class="o">=</span> <span class="mi">21</span>
<span class="o">=========================</span>
      <span class="n">edge</span>   <span class="nb">bin</span>    <span class="n">omega</span>
<span class="o">-------------------------</span>
    <span class="n">Galago</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
    <span class="n">Galago</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
    <span class="n">Galago</span>    <span class="mi">2</span><span class="n">a</span>     <span class="mf">0.00</span>
    <span class="n">Galago</span>    <span class="mi">2</span><span class="n">b</span>     <span class="mf">1.00</span>
 <span class="n">HowlerMon</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
 <span class="n">HowlerMon</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
 <span class="n">HowlerMon</span>    <span class="mi">2</span><span class="n">a</span>     <span class="mf">0.00</span>
 <span class="n">HowlerMon</span>    <span class="mi">2</span><span class="n">b</span>     <span class="mf">1.00</span>
    <span class="n">Rhesus</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
    <span class="n">Rhesus</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
    <span class="n">Rhesus</span>    <span class="mi">2</span><span class="n">a</span>     <span class="mf">0.00</span>
    <span class="n">Rhesus</span>    <span class="mi">2</span><span class="n">b</span>     <span class="mf">1.00</span>
 <span class="n">Orangutan</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
 <span class="n">Orangutan</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
 <span class="n">Orangutan</span>    <span class="mi">2</span><span class="n">a</span>     <span class="mf">0.00</span>
 <span class="n">Orangutan</span>    <span class="mi">2</span><span class="n">b</span>     <span class="mf">1.00</span>
   <span class="n">Gorilla</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
   <span class="n">Gorilla</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
   <span class="n">Gorilla</span>    <span class="mi">2</span><span class="n">a</span>     <span class="mf">0.00</span>
   <span class="n">Gorilla</span>    <span class="mi">2</span><span class="n">b</span>     <span class="mf">1.00</span>
     <span class="n">Human</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
     <span class="n">Human</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
     <span class="n">Human</span>    <span class="mi">2</span><span class="n">a</span>   <span class="mf">100.00</span>
     <span class="n">Human</span>    <span class="mi">2</span><span class="n">b</span>   <span class="mf">100.00</span>
<span class="n">Chimpanzee</span>     <span class="mi">0</span>     <span class="mf">0.00</span>
<span class="n">Chimpanzee</span>     <span class="mi">1</span>     <span class="mf">1.00</span>
<span class="n">Chimpanzee</span>    <span class="mi">2</span><span class="n">a</span>   <span class="mf">100.00</span>
<span class="n">Chimpanzee</span>    <span class="mi">2</span><span class="n">b</span>   <span class="mf">100.00</span><span class="o">...</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.3.<br/>
    </p>
  </div>
</footer>
  </body>
</html>