<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Collections and Alignments &#8212; cogent3  documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Annotations" href="annotations.html" />
    <link rel="prev" title="Protein sequences" href="protein_sequences.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          cogent3</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../install.html">Install</a></li>
                <li><a href="https://cogent3.org/docs">Docs</a></li>
                <li><a href="https://github.com/cogent3/cogent3">Code</a></li>
                <li><a href="https://cogent3.org/cogent3/draw/index.html">Gallery</a></li>
                <li><a href="../projects.html">Projects Using</a></li>
                <li><a href="https://github.com/cogent3/c3dev/wiki/Code-of-Conduct">Code of Conduct</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">The Library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Collections and Alignments</a><ul>
<li><a class="reference internal" href="#basic-collection-objects">Basic Collection objects</a><ul>
<li><a class="reference internal" href="#constructing-a-sequencecollection-or-alignment-object-from-strings">Constructing a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from strings</a></li>
<li><a class="reference internal" href="#constructing-a-arrayalignment-using-make-aligned-seqs">Constructing a <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> using <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code></a></li>
<li><a class="reference internal" href="#converting-a-sequencecollection-to-fasta-format">Converting a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> to FASTA format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a><ul>
<li><a class="reference internal" href="#appending-the-sequences">Appending the sequences</a></li>
<li><a class="reference internal" href="#inserting-the-sequences">Inserting the sequences</a></li>
<li><a class="reference internal" href="#inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence">Inserting sequence(s) based on their alignment to a reference sequence</a><ul>
<li><a class="reference internal" href="#removing-all-columns-with-gaps-in-a-named-sequence">Removing all columns with gaps in a named sequence</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a><ul>
<li><a class="reference internal" href="#accessing-individual-sequences-from-a-collection-or-alignment-by-name">Accessing individual sequences from a collection or alignment by name</a></li>
<li><a class="reference internal" href="#slice-the-sequences-from-an-alignment-like-a-list">Slice the sequences from an alignment like a list</a></li>
<li><a class="reference internal" href="#getting-a-subset-of-sequences-from-the-alignment">Getting a subset of sequences from the alignment</a></li>
</ul>
</li>
<li><a class="reference internal" href="#alignments">Alignments</a><ul>
<li><a class="reference internal" href="#creating-an-alignment-object-from-a-sequencecollection">Creating an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code></a></li>
<li><a class="reference internal" href="#convert-alignment-to-dna-rna-or-protein-moltypes">Convert alignment to DNA, RNA or PROTEIN moltypes</a></li>
<li><a class="reference internal" href="#handling-gaps">Handling gaps</a><ul>
<li><a class="reference internal" href="#remove-all-gaps-from-an-alignment-in-fasta-format">Remove all gaps from an alignment in FASTA format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#writing-sequences-to-file">Writing sequences to file</a></li>
<li><a class="reference internal" href="#converting-an-alignment-to-fasta-format">Converting an alignment to FASTA format</a></li>
<li><a class="reference internal" href="#converting-an-alignment-into-phylip-format">Converting an alignment into Phylip format</a></li>
<li><a class="reference internal" href="#converting-an-alignment-to-a-list-of-strings">Converting an alignment to a list of strings</a></li>
</ul>
</li>
<li><a class="reference internal" href="#slicing-an-alignment">Slicing an alignment</a><ul>
<li><a class="reference internal" href="#by-rows-sequences">By rows (sequences)</a></li>
<li><a class="reference internal" href="#getting-a-single-column-from-an-alignment">Getting a single column from an alignment</a></li>
<li><a class="reference internal" href="#getting-a-region-of-contiguous-columns">Getting a region of contiguous columns</a></li>
<li><a class="reference internal" href="#iterating-over-alignment-positions">Iterating over alignment positions</a></li>
<li><a class="reference internal" href="#getting-codon-3rd-positions-from-alignment">Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
<li><a class="reference internal" href="#getting-codon-3rd-positions-from-arrayalignment">Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code></a></li>
<li><a class="reference internal" href="#filtering-positions">Filtering positions</a><ul>
<li><a class="reference internal" href="#trim-terminal-stop-codons">Trim terminal stop codons</a></li>
<li><a class="reference internal" href="#eliminating-columns-with-non-nucleotide-characters">Eliminating columns with non-nucleotide characters</a></li>
<li><a class="reference internal" href="#getting-all-variable-positions-from-an-alignment">Getting all variable positions from an alignment</a></li>
<li><a class="reference internal" href="#getting-all-constant-positions-from-an-alignment">Getting all constant positions from an alignment</a></li>
<li><a class="reference internal" href="#getting-all-variable-codons-from-an-alignment">Getting all variable codons from an alignment</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filtering-sequences">Filtering sequences</a><ul>
<li><a class="reference internal" href="#extracting-sequences-by-sequence-identifier-into-a-new-alignment-object">Extracting sequences by sequence identifier into a new alignment object</a></li>
<li><a class="reference internal" href="#extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object">Extracting sequences using an arbitrary function into a new alignment object</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#computing-alignment-statistics">Computing alignment statistics</a><ul>
<li><a class="reference internal" href="#getting-motif-counts">Getting motif counts</a></li>
<li><a class="reference internal" href="#computing-motif-probabilities-from-an-alignment">Computing motif probabilities from an alignment</a></li>
<li><a class="reference internal" href="#working-with-alignment-gaps">Working with alignment gaps</a><ul>
<li><a class="reference internal" href="#filtering-extracted-columns-for-the-gap-character">Filtering extracted columns for the gap character</a></li>
<li><a class="reference internal" href="#calculating-the-gap-fraction">Calculating the gap fraction</a></li>
<li><a class="reference internal" href="#extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps">Extracting maps of aligned to unaligned positions (i.e., gap maps)</a></li>
<li><a class="reference internal" href="#filtering-alignments-based-on-gaps">Filtering alignments based on gaps</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="protein_sequences.html" title="Previous Chapter: Protein sequences"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Protein sequences</span>
    </a>
  </li>
  <li>
    <a href="annotations.html" title="Next Chapter: Annotations"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Annotations &raquo;</span>
    </a>
  </li>
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="collections-and-alignments">
<h1>Collections and Alignments<a class="headerlink" href="#collections-and-alignments" title="Permalink to this headline">¶</a></h1>
<p>For loading collections of unaligned or aligned sequences see <a class="reference internal" href="loading_sequences.html#load-seqs"><span class="std std-ref">Loading nucleotide, protein sequences</span></a>.</p>
<div class="section" id="basic-collection-objects">
<h2>Basic Collection objects<a class="headerlink" href="#basic-collection-objects" title="Permalink to this headline">¶</a></h2>
<div class="section" id="constructing-a-sequencecollection-or-alignment-object-from-strings">
<h3>Constructing a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from strings<a class="headerlink" href="#constructing-a-sequencecollection-or-alignment-object-from-strings" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="constructing-a-arrayalignment-using-make-aligned-seqs">
<h3>Constructing a <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> using <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code><a class="headerlink" href="#constructing-a-arrayalignment-using-make-aligned-seqs" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="converting-a-sequencecollection-to-fasta-format">
<h3>Converting a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> to FASTA format<a class="headerlink" href="#converting-a-sequencecollection-to-fasta-format" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="adding-new-sequences-to-an-existing-collection-or-alignment">
<h2>Adding new sequences to an existing collection or alignment<a class="headerlink" href="#adding-new-sequences-to-an-existing-collection-or-alignment" title="Permalink to this headline">¶</a></h2>
<p>New sequences can be either appended or inserted using the <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> method. More than one sequence can be added at the same time. Note that <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> does not modify the existing collection/alignment, it creates a new one.</p>
<div class="section" id="appending-the-sequences">
<h3>Appending the sequences<a class="headerlink" href="#appending-the-sequences" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> without additional parameters will append the sequences to the end of the collection/alignment.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The order is not preserved if you use <code class="docutils literal notranslate"><span class="pre">to_fasta</span></code> method, which sorts sequences by name.</p>
</div>
</div>
<div class="section" id="inserting-the-sequences">
<h3>Inserting the sequences<a class="headerlink" href="#inserting-the-sequences" title="Permalink to this headline">¶</a></h3>
<p>Sequences can be inserted into an alignment at the specified position using either the <code class="docutils literal notranslate"><span class="pre">before_name</span></code> or <code class="docutils literal notranslate"><span class="pre">after_name</span></code> arguments.</p>
</div>
<div class="section" id="inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence">
<h3>Inserting sequence(s) based on their alignment to a reference sequence<a class="headerlink" href="#inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence" title="Permalink to this headline">¶</a></h3>
<p>Already aligned sequences can be added to an existing <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object and aligned at the same time using the <code class="docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code> method. The alignment is performed based on their alignment to a reference sequence (which must be present in both alignments). The method assumes the first sequence in <code class="docutils literal notranslate"><span class="pre">ref_aln.names[0]</span></code> is the reference.</p>
<p><code class="docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code> has the same arguments as <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> so <code class="docutils literal notranslate"><span class="pre">before_name</span></code> and <code class="docutils literal notranslate"><span class="pre">after_name</span></code> can be used to insert the new sequences at the desired position.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method does not work with the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> class.</p>
</div>
<div class="section" id="removing-all-columns-with-gaps-in-a-named-sequence">
<h4>Removing all columns with gaps in a named sequence<a class="headerlink" href="#removing-all-columns-with-gaps-in-a-named-sequence" title="Permalink to this headline">¶</a></h4>
</div>
</div>
</div>
<div class="section" id="the-elements-of-a-collection-or-alignment">
<h2>The elements of a collection or alignment<a class="headerlink" href="#the-elements-of-a-collection-or-alignment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="accessing-individual-sequences-from-a-collection-or-alignment-by-name">
<h3>Accessing individual sequences from a collection or alignment by name<a class="headerlink" href="#accessing-individual-sequences-from-a-collection-or-alignment-by-name" title="Permalink to this headline">¶</a></h3>
<p>Using the <code class="docutils literal notranslate"><span class="pre">get_seq</span></code> method allows for extracting an unaligned sequence from a collection or alignment by name.</p>
<p>Alternatively, if you want to extract the aligned (i.e., gapped) sequence from an alignment, you can use <code class="docutils literal notranslate"><span class="pre">get_gapped_seq</span></code>.</p>
<p>To see the names of the sequences in a sequence collection, you can use either the <code class="docutils literal notranslate"><span class="pre">Names</span></code> attribute or <code class="docutils literal notranslate"><span class="pre">get_seq_names</span></code> method.</p>
</div>
<div class="section" id="slice-the-sequences-from-an-alignment-like-a-list">
<h3>Slice the sequences from an alignment like a list<a class="headerlink" href="#slice-the-sequences-from-an-alignment-like-a-list" title="Permalink to this headline">¶</a></h3>
<p>The usual approach is to access a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object as a dictionary, obtaining the individual sequences using the titles as “keys” (above).  However, one can also iterate through the collection like a list.</p>
</div>
<div class="section" id="getting-a-subset-of-sequences-from-the-alignment">
<h3>Getting a subset of sequences from the alignment<a class="headerlink" href="#getting-a-subset-of-sequences-from-the-alignment" title="Permalink to this headline">¶</a></h3>
<p>Note, if you set <code class="docutils literal notranslate"><span class="pre">array_align=False</span></code>, then the subset contain references to the original sequences, not copies.</p>
</div>
</div>
<div class="section" id="alignments">
<h2>Alignments<a class="headerlink" href="#alignments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="creating-an-alignment-object-from-a-sequencecollection">
<h3>Creating an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code><a class="headerlink" href="#creating-an-alignment-object-from-a-sequencecollection" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="convert-alignment-to-dna-rna-or-protein-moltypes">
<h3>Convert alignment to DNA, RNA or PROTEIN moltypes<a class="headerlink" href="#convert-alignment-to-dna-rna-or-protein-moltypes" title="Permalink to this headline">¶</a></h3>
<p>This is useful if you’ve loaded a sequence alignment without specifying the moltype and later need to convert it.</p>
<p>To RNA</p>
<p>To PROTEIN</p>
</div>
<div class="section" id="handling-gaps">
<h3>Handling gaps<a class="headerlink" href="#handling-gaps" title="Permalink to this headline">¶</a></h3>
<div class="section" id="remove-all-gaps-from-an-alignment-in-fasta-format">
<h4>Remove all gaps from an alignment in FASTA format<a class="headerlink" href="#remove-all-gaps-from-an-alignment-in-fasta-format" title="Permalink to this headline">¶</a></h4>
<p>This necessarily returns a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code>.</p>
</div>
</div>
<div class="section" id="writing-sequences-to-file">
<h3>Writing sequences to file<a class="headerlink" href="#writing-sequences-to-file" title="Permalink to this headline">¶</a></h3>
<p>Both collection and alignment objects have a <code class="docutils literal notranslate"><span class="pre">write</span></code> method. The output format is inferred from the filename suffix,</p>
<p>or by the <code class="docutils literal notranslate"><span class="pre">format</span></code> argument.</p>
</div>
<div class="section" id="converting-an-alignment-to-fasta-format">
<h3>Converting an alignment to FASTA format<a class="headerlink" href="#converting-an-alignment-to-fasta-format" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="converting-an-alignment-into-phylip-format">
<h3>Converting an alignment into Phylip format<a class="headerlink" href="#converting-an-alignment-into-phylip-format" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="converting-an-alignment-to-a-list-of-strings">
<h3>Converting an alignment to a list of strings<a class="headerlink" href="#converting-an-alignment-to-a-list-of-strings" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="slicing-an-alignment">
<h2>Slicing an alignment<a class="headerlink" href="#slicing-an-alignment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="by-rows-sequences">
<h3>By rows (sequences)<a class="headerlink" href="#by-rows-sequences" title="Permalink to this headline">¶</a></h3>
<p>An <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> can be sliced</p>
<p>but a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> cannot be sliced</p>
</div>
<div class="section" id="getting-a-single-column-from-an-alignment">
<h3>Getting a single column from an alignment<a class="headerlink" href="#getting-a-single-column-from-an-alignment" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="getting-a-region-of-contiguous-columns">
<h3>Getting a region of contiguous columns<a class="headerlink" href="#getting-a-region-of-contiguous-columns" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="iterating-over-alignment-positions">
<h3>Iterating over alignment positions<a class="headerlink" href="#iterating-over-alignment-positions" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="getting-codon-3rd-positions-from-alignment">
<h3>Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#getting-codon-3rd-positions-from-alignment" title="Permalink to this headline">¶</a></h3>
<p>We’ll do this by specifying the position indices of interest, creating a sequence <code class="docutils literal notranslate"><span class="pre">Feature</span></code> and using that to extract the positions.</p>
</div>
<div class="section" id="getting-codon-3rd-positions-from-arrayalignment">
<h3>Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code><a class="headerlink" href="#getting-codon-3rd-positions-from-arrayalignment" title="Permalink to this headline">¶</a></h3>
<p>We can use more conventional slice notation in this instance. Note, because Python counts from 0, the 3rd position starts at index 2.</p>
</div>
<div class="section" id="filtering-positions">
<span id="filter-positions"></span><h3>Filtering positions<a class="headerlink" href="#filtering-positions" title="Permalink to this headline">¶</a></h3>
<div class="section" id="trim-terminal-stop-codons">
<h4>Trim terminal stop codons<a class="headerlink" href="#trim-terminal-stop-codons" title="Permalink to this headline">¶</a></h4>
<p>For evolutionary analyses that use codon models we need to exclude terminating stop codons. For the case where the sequences are all of length divisible by 3.</p>
<p>If the alignment contains sequences not divisible by 3, use the <code class="docutils literal notranslate"><span class="pre">allow_partial</span></code> argument.</p>
</div>
<div class="section" id="eliminating-columns-with-non-nucleotide-characters">
<h4>Eliminating columns with non-nucleotide characters<a class="headerlink" href="#eliminating-columns-with-non-nucleotide-characters" title="Permalink to this headline">¶</a></h4>
<p>We sometimes want to eliminate ambiguous or gap data from our alignments. We show how to exclude alignment columns by the characters they contain. In the first instance we do this just for single nucleotide columns, then for trinucleotides (equivalent for handling codons). Both are done using the <code class="docutils literal notranslate"><span class="pre">no_degenerates</span></code> method.</p>
<p>We apply to nucleotides,</p>
<p>Applying the same filter to trinucleotides (specified by setting <code class="docutils literal notranslate"><span class="pre">motif_length=3</span></code>).</p>
</div>
<div class="section" id="getting-all-variable-positions-from-an-alignment">
<h4>Getting all variable positions from an alignment<a class="headerlink" href="#getting-all-variable-positions-from-an-alignment" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="getting-all-constant-positions-from-an-alignment">
<h4>Getting all constant positions from an alignment<a class="headerlink" href="#getting-all-constant-positions-from-an-alignment" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="getting-all-variable-codons-from-an-alignment">
<h4>Getting all variable codons from an alignment<a class="headerlink" href="#getting-all-variable-codons-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<p>This is done using the <code class="docutils literal notranslate"><span class="pre">filtered</span></code> method using the <code class="docutils literal notranslate"><span class="pre">motif_length</span></code> argument. We demonstrate this first for the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>.</p>
<p>Then for the standard <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> by first converting the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>.</p>
</div>
</div>
<div class="section" id="filtering-sequences">
<h3>Filtering sequences<a class="headerlink" href="#filtering-sequences" title="Permalink to this headline">¶</a></h3>
<div class="section" id="extracting-sequences-by-sequence-identifier-into-a-new-alignment-object">
<h4>Extracting sequences by sequence identifier into a new alignment object<a class="headerlink" href="#extracting-sequences-by-sequence-identifier-into-a-new-alignment-object" title="Permalink to this headline">¶</a></h4>
<p>You can use <code class="docutils literal notranslate"><span class="pre">take_seqs</span></code> to extract some sequences by sequence identifier from an alignment to a new alignment object:</p>
<p>Alternatively, you can extract only the sequences which are not specified by passing <code class="docutils literal notranslate"><span class="pre">negate=True</span></code>:</p>
</div>
<div class="section" id="extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object">
<h4>Extracting sequences using an arbitrary function into a new alignment object<a class="headerlink" href="#extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object" title="Permalink to this headline">¶</a></h4>
<p>You can use <code class="docutils literal notranslate"><span class="pre">take_seqs_if</span></code> to extract sequences into a new alignment object based on whether an arbitrary function applied to the sequence evaluates to True. For example, to extract sequences which don’t contain any N bases you could do the following:</p>
<p>You can additionally get the sequences where the provided function evaluates to False:</p>
</div>
</div>
</div>
<div class="section" id="computing-alignment-statistics">
<h2>Computing alignment statistics<a class="headerlink" href="#computing-alignment-statistics" title="Permalink to this headline">¶</a></h2>
<div class="section" id="getting-motif-counts">
<h3>Getting motif counts<a class="headerlink" href="#getting-motif-counts" title="Permalink to this headline">¶</a></h3>
<p>We state the motif length we want and whether to allow gap or ambiguous characters. The latter only has meaning for IPUAC character sets (the DNA, RNA or PROTEIN moltypes). We illustrate this for the DNA moltype with motif lengths of 1 and 3.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only the observed motifs are returned, rather than all defined by the alphabet.</p>
</div>
</div>
<div class="section" id="computing-motif-probabilities-from-an-alignment">
<h3>Computing motif probabilities from an alignment<a class="headerlink" href="#computing-motif-probabilities-from-an-alignment" title="Permalink to this headline">¶</a></h3>
<p>The method <code class="docutils literal notranslate"><span class="pre">get_motif_probs</span></code> of <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects returns the probabilities for all motifs of a given length. For individual nucleotides:</p>
<p>For dinucleotides or longer, we need to pass in an <code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> with the appropriate word length. Here is an example with trinucleotides:</p>
<p>The same holds for other arbitrary alphabets, as long as they match the alignment <code class="docutils literal notranslate"><span class="pre">MolType</span></code>.</p>
<p>Some calculations in <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> require all non-zero values in the motif probabilities, in which case we use a pseudo-count. We illustrate that here with a simple example where T is missing. Without the pseudo-count, the frequency of T is 0.0, with the pseudo-count defined as 1e-6 then the frequency of T will be slightly less than 1e-6.</p>
<p>It is important to notice that motif probabilities are computed by treating sequences as non-overlapping tuples. Below is a very simple pair of identical sequences where there are clearly 2 ‘AA’ dinucleotides per sequence but only the first one is ‘in-frame’ (frame width = 2).</p>
<p>We then create a dinucleotide <code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> object and use this to get dinucleotide probabilities. These frequencies are determined by breaking each aligned sequence up into non-overlapping dinucleotides and then doing a count. The expected value for the ‘AA’ dinucleotide in this case will be 2/8 = 0.25.</p>
<p>What about counting the total incidence of dinucleotides including those not in-frame?  A naive application of the Python string object’s count method will not work as desired either because it “returns the number of non-overlapping occurrences”.</p>
<p>To count all occurrences of a given dinucleotide in a DNA sequence, one could use a standard Python approach such as list comprehension:</p>
</div>
<div class="section" id="working-with-alignment-gaps">
<h3>Working with alignment gaps<a class="headerlink" href="#working-with-alignment-gaps" title="Permalink to this headline">¶</a></h3>
<div class="section" id="filtering-extracted-columns-for-the-gap-character">
<h4>Filtering extracted columns for the gap character<a class="headerlink" href="#filtering-extracted-columns-for-the-gap-character" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="calculating-the-gap-fraction">
<h4>Calculating the gap fraction<a class="headerlink" href="#calculating-the-gap-fraction" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps">
<h4>Extracting maps of aligned to unaligned positions (i.e., gap maps)<a class="headerlink" href="#extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps" title="Permalink to this headline">¶</a></h4>
<p>It’s often important to know how an alignment position relates to a position in one or more of the sequences in the alignment. The <code class="docutils literal notranslate"><span class="pre">gap_maps</span></code> method of the individual sequences is useful for this. To get a map of sequence to alignment positions for a specific sequence in your alignment, do the following:</p>
<p>It’s now possible to look up positions in the <code class="docutils literal notranslate"><span class="pre">seq1</span></code>, and find out what they map to in the alignment:</p>
<p>This tells us that in position 3 in <code class="docutils literal notranslate"><span class="pre">seq1</span></code> corresponds to position 3 in <code class="docutils literal notranslate"><span class="pre">aln</span></code>, and that position 8 in <code class="docutils literal notranslate"><span class="pre">seq1</span></code> corresponds to position 9 in <code class="docutils literal notranslate"><span class="pre">aln</span></code>.</p>
<p>Notice that we grabbed the first result from the call to <code class="docutils literal notranslate"><span class="pre">gap_maps</span></code>. This is the sequence position to alignment position map. The second value returned is the alignment position to sequence position map, so if you want to find out what sequence positions the alignment positions correspond to (opposed to what alignment positions the sequence positions correspond to) for a given sequence, you would take the following steps:</p>
<p>If an alignment position is a gap, and therefore has no corresponding sequence position, you’ll get a <code class="docutils literal notranslate"><span class="pre">KeyError</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The first position in alignments and sequences is always numbered position 0.</p>
</div>
</div>
<div class="section" id="filtering-alignments-based-on-gaps">
<h4>Filtering alignments based on gaps<a class="headerlink" href="#filtering-alignments-based-on-gaps" title="Permalink to this headline">¶</a></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An alternate, computationally faster, approach to removing gaps is to use the <code class="docutils literal notranslate"><span class="pre">filtered</span></code> method as discussed in <a class="reference internal" href="#filter-positions"><span class="std std-ref">Filtering positions</span></a>.</p>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">omit_gap_runs</span></code> method can be applied to remove long stretches of gaps in an alignment. In the following example, we remove sequences that have more than two adjacent gaps anywhere in the aligned sequence.</p>
<p>If instead, we just wanted to remove positions from the alignment which are gaps in more than a certain percentage of the sequences, we could use the <code class="docutils literal notranslate"><span class="pre">omit_gap_pos</span></code> function. For example:</p>
<p>You’ll notice that the 4th and 7th columns of the alignment have been removed because they contained 66% gaps – more than the allowed 40%.</p>
<p>If you wanted to remove sequences which contain more than a certain percent gap characters, you could use the <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code> method. This is commonly applied to filter partial sequences from an alignment.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGAA------&#39;</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="p">(</span><span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;ATG-AGTGATG&#39;</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="p">(</span><span class="s1">&#39;seq3&#39;</span><span class="p">,</span> <span class="s1">&#39;AT--AG-GATG&#39;</span><span class="p">)],</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filtered_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_seqs</span><span class="p">(</span><span class="mf">0.50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">filtered_aln</span><span class="p">)</span>  
<span class="go">&gt;seq2</span>
<span class="go">ATG-AGTGATG</span>
<span class="go">&gt;seq3</span>
<span class="go">AT--AG-GATG</span>
</pre></div>
</div>
<p>Note that following this call to <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code>, the 4th column of <code class="docutils literal notranslate"><span class="pre">filtered_aln</span></code> is 100% gaps. This is generally not desirable, so a call to <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code> is frequently followed with a call to <code class="docutils literal notranslate"><span class="pre">omit_gap_pos</span></code> with no parameters – this defaults to removing positions which are all gaps:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">filtered_aln</span><span class="o">.</span><span class="n">omit_gap_pos</span><span class="p">())</span>  
<span class="go">&gt;seq2</span>
<span class="go">ATGAGTGATG</span>
<span class="go">&gt;seq3</span>
<span class="go">AT-AG-GATG</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.3.<br/>
    </p>
  </div>
</footer>
  </body>
</html>